compiler/rustc_session/src/session.rs

```diff
@@ -250,14 +251,9 @@ impl Session {
         self.miri_unleashed_features.lock().push((span, feature_gate));
     }
 
-    pub fn local_crate_source_file(&self) -> Option<PathBuf> {
+    pub fn local_crate_source_file(&self) -> Option<RealFileName> {
         let path = self.io.input.opt_path()?;
-        // FIXME: The remap path scope should probably not be hardcoded.
-        if self.should_prefer_remapped(RemapPathScopeComponents::DEBUGINFO) {
-            Some(self.opts.file_path_mapping().map_prefix(path).0.into_owned())
-        } else {
-            Some(path.to_path_buf())
-        }
+        Some(RealFileName::LocalPath(path.to_path_buf()))
     }
```

Comment:
The method needs to apply the remapping. Otherwise `.for_scope` won't be able to return the remapped version because on the local file system path is available in the `RealFileName` value.

I suggest adding a helper method to [`FilePathMapping`](https://github.com/rust-lang/rust/blob/47dd709bedda8127e8daec33327e0a9d0cdae845/compiler/rustc_span/src/source_map.rs#L1047) to make creating `RealFileName`s easy:
```rust
pub fn to_real_filename<'a>(&'a self, local_path: impl Into<Cow<'a, Path>>) -> RealFileName {
    let local_path = local_path.into();
    if let (remapped_path, true) = self.map_prefix(&*local_path) {
        RealFileName::Remapped {
            virtual_name: remapped_path.into_owned(),
            local_path: Some(local_path.into_owned()),
        }
    } else {
        RealFileName::LocalPath(local_path.into_owned())
    }
}
```
Then you can implement `local_crate_source_file` as:
```rust
pub fn local_crate_source_file(&self) -> Option<RealFileName> {
    Some(self.source_map().path_mapping().to_real_filename(self.io.input.opt_path()?))
}
```

The `FilePathMapping::to_real_filename` helper can then also be used to create `Session::working_dir`.

Reply:
Done and done.

*Btw, thanks you for the help and sorry this extra unnecessary work, I should have done the remapping from the start; I really shouldn't do 3 things at the same time 😅.*

Reply:
> I really shouldn't do 3 things at the same time 😅

I can relate to that one 😉